src/circuit/qc_keyagg.rs

```diff
@@ -141,42 +192,169 @@ where
+    }
+}
+
+impl<E> VerKeyVar<E> for EmulatedTEPointVariable<E>
+where
+    E: PrimeField,
+{
+    type KeyType = TEPoint<E>;
+
+    fn native_vars(&self) -> Vec<Variable> {
+        let mut ret = self.0.native_vars();
+        ret.append(&mut self.1.native_vars());
+        ret
+    }
+
+    fn aggregate_with_selectors<F>(
+        circuit: &mut PlonkCircuit<F>,
+        keys: &[Self],
+        selectors: &[BoolVar],
+        coef: E,
+    ) -> Result<Self, CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>,
+    {
+        let neutral_point = Self::KeyType::default();
+        let emulated_neutral_point_var =
+            circuit.create_constant_emulated_te_point_variable(neutral_point)?;
+        let mut agg_key_var = emulated_neutral_point_var.clone();
+        for (key, &bit) in keys.iter().zip(selectors.iter()) {
+            let point_var = circuit.binary_emulated_te_point_vars_select(
+                bit,
+                &emulated_neutral_point_var,
+                key,
+            )?;
+            agg_key_var = circuit.emulated_te_ecc_add::<E>(&agg_key_var, &point_var, coef)?;
+        }
+        Ok(agg_key_var)
+    }
+
+    fn is_equal<F>(
+        circuit: &mut PlonkCircuit<F>,
+        p0: &Self,
+        p1: &Self,
+    ) -> Result<BoolVar, CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>,
+    {
+        circuit.is_emulated_te_point_equal(p0, p1)
+    }
+
+    fn enforce_equal<F>(
+        circuit: &mut PlonkCircuit<F>,
+        p0: &Self,
+        p1: &Self,
+    ) -> Result<(), CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>,
+    {
+        circuit.enforce_emulated_te_point_equal(p0, p1)
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
     use ark_bls12_377::{g1::Config as Param377, Fq as Fq377};
-    use ark_bn254::Fr as Fr254;
+    use ark_bn254::{g1::Config as Param254, Fq as Fq254, Fr as Fr254};
+    use ark_ec::{
+        short_weierstrass::{Projective, SWCurveConfig},
+        CurveGroup,
+    };
     use ark_ff::MontFp;
-    use ark_std::{vec::Vec, UniformRand};
-    use jf_primitives::rescue::sponge::RescueCRHF;
+    use ark_std::{vec::Vec, UniformRand, Zero};
     use jf_relation::{
-        errors::CircuitError,
-        gadgets::{ecc::SWToTEConParam, from_emulated_field},
-        Circuit, PlonkCircuit, Variable,
+        errors::CircuitError, gadgets::ecc::SWToTEConParam, Circuit, PlonkCircuit, Variable,
     };
 
     #[test]
-    fn test_vk_aggregate_circuit() -> Result<(), CircuitError> {
-        let d_ecc : Fq377 = MontFp!("122268283598675559488486339158635529096981886914877139579534153582033676785385790730042363341236035746924960903179");
-        test_vk_aggregate_circuit_helper::<Fq377, Fr254, Param377>(d_ecc)
+    fn test_vk_aggregate_sw_circuit() -> Result<(), CircuitError> {
+        let a_ecc = Fq377::zero();
+        test_vk_aggregate_sw_circuit_helper::<Fq377, Fr254, Param377>(a_ecc)?;
```

Comment:
Why are we using Fr254 rather than Fr377 here?

Reply:
We only have emulation config for native field `Fr254`. IMO that's sufficient, nothing special for `Fr377`.